ambry-replication/src/main/java/com.github.ambry.replication/CloudToStoreReplicationManager.java

```diff
@@ -0,0 +1,333 @@
+   * @param transformerClassName name of the class to transform and validate replication messages.
+   * @param vcrClusterSpectator {@link ClusterSpectator} object to get changes in vcr cluster map.
+   * @param clusterParticipant {@link ClusterParticipant} object to get changes in partition state of partitions on datanodes.
+   * @throws ReplicationException
+   */
+  public CloudToStoreReplicationManager(ReplicationConfig replicationConfig, ClusterMapConfig clusterMapConfig,
+      StoreConfig storeConfig, StoreManager storeManager, StoreKeyFactory storeKeyFactory, ClusterMap clusterMap,
+      ScheduledExecutorService scheduler, DataNodeId currentNode, ConnectionPool connectionPool,
+      MetricRegistry metricRegistry, NotificationSystem requestNotification,
+      StoreKeyConverterFactory storeKeyConverterFactory, String transformerClassName,
+      ClusterSpectator vcrClusterSpectator, ClusterParticipant clusterParticipant) throws ReplicationException {
+    super(replicationConfig, clusterMapConfig, storeKeyFactory, clusterMap, scheduler, currentNode,
+        Collections.emptyList(), connectionPool, metricRegistry, requestNotification, storeKeyConverterFactory,
+        transformerClassName);
+    this.clusterMapConfig = clusterMapConfig;
+    this.storeConfig = storeConfig;
+    this.storeManager = storeManager;
+    this.vcrClusterSpectator = vcrClusterSpectator;
+    this.clusterParticipant = clusterParticipant;
+    this.instanceNameToCloudDataNode = new ConcurrentHashMap<>();
+    this.vcrNodes = new ConcurrentSkipListSet<>();
+    this.persistor =
+        new DiskTokenPersistor(cloudReplicaTokenFileName, mountPathToPartitionInfos, replicationMetrics, clusterMap,
+            tokenHelper);
+    this.localPartitionNameToPartition = mapPartitionNameToPartition(clusterMap, currentNode);
+  }
+
+  private ConcurrentHashMap<String, PartitionId> mapPartitionNameToPartition(ClusterMap clusterMap,
+      DataNodeId localNode) {
+    List<? extends ReplicaId> localReplicas = clusterMap.getReplicaIds(localNode);
+    return localReplicas.stream()
+        .collect(Collectors.toMap(replicaId -> replicaId.getPartitionId().toPathString(), ReplicaId::getPartitionId,
+            (e1, e2) -> e2, ConcurrentHashMap::new));
+  }
+
+  @Override
+  public void start() throws ReplicationException {
+    // Add listener for vcr instance config changes
+    vcrClusterSpectator.registerInstanceConfigChangeListener(new InstanceConfigChangeListener() {
+      @Override
+      public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+        ConcurrentSkipListSet<CloudDataNode> newVcrNodes = new ConcurrentSkipListSet<>();
+        ConcurrentHashMap<String, CloudDataNode> newInstanceNameToCloudDataNode = new ConcurrentHashMap<>();
+        Set<CloudDataNode> removedNodes = new HashSet<>(vcrNodes);
+        synchronized (notificationLock) {
+          for (InstanceConfig instanceConfig : instanceConfigs) {
```

Comment:
Please add some comments here, including why you need to remove and then add the same partitions.

Reply:
fixed.